### DISCUSSION TITLE
E1: admin password not working - can I use user password?

### DISCUSSION THREAD

**Discussion by jsn081** [GerryDazoo/Slinger#642] [General]
So I have been gifted a 500 and a 350 (on different sling accounts).  At the time I received them I could download the encrypted passwords, but couldn't get the plaintext passwords.  Note that I've been using Slinger happily for a year both by itself and integrated to Plex on my own personal 350 that I have the plaintext password for.  I'm now looking to see if I can do multi-tuner.

I haven't touched the 350 yet, but the 500 I'm trying to get to work was in an account with multiple boxes.  So the admin password is listed under a different box than the one I'm trying to use.  It doesn't feel like this should matter, but the password is not working.  I also attempted the user password and "admin".  My questions are:

- Has there ever been a way to decrypt the passwords or brute force them?
- Can I use the user password as well, or does the software only take the admin password?
- If I reset the 350 or 500 and use admin, will that work permanently?  I've heard suggestion that it resets to the old password on power failure and that you lose IR control.

Note that losing IR control isn't ideal but isn't the worst in my case.  I'm trying to control TiVos, and they have the ability to be controlled via TCP commands.  Once I had video coming out of my Slingbox I was going to attempt to modify Slinger to do channel changes via TiVo network remote - since this would be faster and more reliable than IR blasting.  I downloaded the protocol document for it.

Log snippet (in debug mode) and auth response posted below.  Notes from the log: Jasonsling is a 350 that's working, DanSling500 is not.  The finderID in discovery mode (from v3.08f) matches the one in the auth response so I know I have the right box.  I tried to initiate streams on both boxes.  Note in the auth response that the admin password hash seems shorter than normal??

Thanks in advance.

/var/sling.log:
```
Version : 4.01 Running on Linux-6.2.0-39-generic-x86_64-with-glibc2.37 pid= 25976 slingbox_server.py
Using config file config.ini
Connection Manager Running on port 8081 with 4 max streams using URL slinger.
BOXES [('sb1', '350'), ('sb2', '500')]
BOX URL /slinger/350
Building page for 350
Reading Custom Remote definition from Tivo.Premiere.Remote.txt
BOX URL Streamer Running:  4 /slinger/500config.ini 350 
Building page for350  8081500 
8388608Reading Custom Remote definition from 
Tivo.Premiere.Remote.txt
JasonSling Checking for slingbox at ('10.21.0.89', 5201) 0
starting up on  port 8081 Streamer Running: 8388608 
4 config.ini 500 500 8081 8388608
DanSling500 Checking for slingbox atJasonSling ('10.21.0.89', 5201)  ('10.21.0.90', 5209) OK
0
JasonSling Using slingbox at  ('10.21.0.89', 5201)
Streamer:  JasonSling Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
DanSling500 ('10.21.0.90', 5209) OK
DanSling500 Using slingbox at  ('10.21.0.90', 5209)
Streamer:  DanSling500 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
 * Serving Flask app 'slingbox_server'
 * Debug mode: on
Broadcast Responder Running
Server IP address is 10.21.0.14
01/18/2024, 11:18:24.945   Streaming request from ('127.0.0.1', 50472)  /slinger/350?channel=3 
STREAM=127.0.0.1:50472:3

Slinginfo  "350/500/M1/M2" 12 30 10.21.0.89 5201 3072 10 X7Ekfb32aIjQxZ2
JasonSling Starting Stream for  127.0.0.1:50472
JasonSling Opening Control stream 0x2000 0 0
Connecting... ('10.21.0.89', 5201) Control
VideoParameters: Resolution= 12 FrameRate= 30 VideoBandwidth= 1200 VideoSmoothness= 48 IframeRate= 3 AudioBitRate= 48
Connecting... ('10.21.0.89', 5201) Stream
JasonSling Stream started at 01/18/2024, 11:18:27.329  744 0
01/18/2024, 11:18:27.329  JasonSling Sending Start Channel 3
sending key 11 1
....01/18/2024, 11:19:18.34  JasonSling Stream Terminated for  127.0.0.1:50472
JasonSling Shutting down connections
JasonSling Logging Out
Streamer:  JasonSling Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
01/18/2024, 11:19:18.817   Streaming request from ('127.0.0.1', 58086)  /slinger/350?channel=200 
STREAM=127.0.0.1:58086:200

Slinginfo  "350/500/M1/M2" 12 30 10.21.0.89 5201 3072 10 X7Ekfb32aIjQxZ2
JasonSling Starting Stream for  127.0.0.1:58086
JasonSling Opening Control stream 0x2000 0 0
Connecting... ('10.21.0.89', 5201) Control
VideoParameters: Resolution= 12 FrameRate= 30 VideoBandwidth= 1200 VideoSmoothness= 48 IframeRate= 3 AudioBitRate= 48
Connecting... ('10.21.0.89', 5201) Stream
JasonSling Stream started at 01/18/2024, 11:19:22.212  744 0
01/18/2024, 11:19:22.212  JasonSling Sending Start Channel 200
sending key 10 1
sending key 18 1
sending key 18 1
.....01/18/2024, 11:20:17.080   Streaming request from ('10.21.2.2', 55428)  /slinger/500 
STREAM=10.21.2.2:55428:0

Slinginfo  "350/500/M1/M2" 12 30 10.21.0.90 5209 3072 10 admin
DanSling500 Starting Stream for  10.21.2.2:55428
DanSling500 Opening Control stream 0x2000 0 0
Connecting... ('10.21.0.90', 5209) Control
cmd: 0x67 err: 0x2 8
DanSling500 Error Starting Session. Check your admin password in config.ini file!
DanSling500 ERROR: Slingbox session startup failed.
DanSling500 Logging Out
Streamer:  DanSling500 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
01/18/2024, 11:20:17.407   Streaming request from ('10.21.2.2', 55429)  /slinger/500 
STREAM=10.21.2.2:55429:0

Slinginfo  "350/500/M1/M2" 12 30 10.21.0.90 5209 3072 10 admin
DanSling500 Starting Stream for  10.21.2.2:55429
DanSling500 Opening Control stream 0x2000 0 0
Connecting... ('10.21.0.90', 5209) Control
cmd: 0x67 err: 0x2 8
DanSling500 Error Starting Session. Check your admin password in config.ini file!
DanSling500 ERROR: Slingbox session startup failed.
DanSling500 Logging Out
Streamer:  DanSling500 Waiting for first stream, flushing any IR requests that arrive while not connected to slingbox
01/18/2024, 11:20:17.651   Streaming request from ('10.21.2.2', 55430)  /slinger/500 
STREAM=10.21.2.2:55430:0
```

Auth Response:
```
<finderId>F0F13CCD0B4A1A41B8F39C2DA0491FB4</finderId>
<memberSlingBoxId>1539201</memberSlingBoxId>
<adminPassword>
E1:90d11912d7cbf3a17cf8a77ac157dc029674363e0690a304
</adminPassword>
<productSignature>000008</productSignature>
<displayName>Dan's Slingbox</displayName>

<memberSlingBoxId>2814727</memberSlingBoxId>
<finderId>B56A76BE19FA95CA13A76B32041132DD</finderId>
<productSignature>000012</productSignature>
<userPassword>
E1:47c5fab48582dc9f23638dba951e0a78b53a39aec31d4a3307123e3bc828b71e
</userPassword>
<displayName>Dan's Slingbox</displayName>

<memberSlingBoxId>2749204</memberSlingBoxId>
<finderId>C2888AD4A7CA8C9C30D0899849FC3D96</finderId>
<productSignature>000012</productSignature>
<userPassword>
E1:3c55ea8c15b4b3d523a5fc2bfcf5dd30543f34f3fce98e7a5801fa0da6ecea88
</userPassword>
<displayName>GIBSLING</displayName>

<memberSlingBoxId>2761680</memberSlingBoxId>
<finderId>33DCCF5F50EAF75F3D78BE3E0A31299D</finderId>
<productSignature>000012</productSignature>
<userPassword>
E1:310ed2b7553f04f6de3147c460e99ccc1506355d91e8d243e912a6ccafbf21c4
</userPassword>
<displayName>Control Room</displayName>
```

**Comment by GerryDazoo** [comment]
Well written post with useful information. I'll try to answer your questions.

    Has there ever been a way to decrypt the passwords or brute force them?
The Slinger software will eat an encrypted password just put the appropriate string in your config.ini
i.e. password=E1:90d11912d7cbf3a17cf8a77ac157dc029674363e0690a304
the preceding E1: tells the code it's encrypted. 

    Can I use the user password as well, or does the software only take the admin password?
I only use the admin password but I believe the user password will work as well.

    If I reset the 350 or 500 and use admin, will that work permanently? 
Yes that will work permanently but the Factory reset also wipes out the IR control definitions that were previously configured. But unless you have an identical cable box to the one that had been configured by the original owner the IR functions won't work anyway, so in that respect you're not any worse off. 
I haven't tested this lately but it was possible to install a custom 3rd party remote definition with SlingPlayer. Using the "built in" remote definitions stopped working when the servers went down.

**Reply by jsn081** [reply]
I find it odd that the admin password hash I got is shorter than normal.  Definitely shorter than the other 3 accounts I've extracted.  Have you ran into that before?

Given that the user and admin passwords don't work, is my only option to do a reset?  Or am I missing any other step before I attempt it?  I tried to download old desktop software, but all of the versions just stick on the logo screen and never go further.

**Reply by rbf121** [reply]
Yes, I have seen shorter passwords.  Some of the older slingbox products would like you specify a custom password rather than auto generated one.  Custom passwords could vary in length.  In your case, the shorter password decrypted would be "gcwest81" and it belongs to a Slingbox PRO-HD (product id 8).  

But the slinger software should work with encrypted and decrypted passwords so it shouldn't matter.  I think maybe the passwords you are trying do not match the box.  A 500 has a product signature of 18 which isn't in your list.

**Reply by GerryDazoo** [reply]
rbf121 is correct. I'm wondering if somehow your boxes may have been reset after you recovered the passwords.

**Comment by jsn081** [comment]
> A 500 has a product signature of 18 which isn't in your list.

The model of the box says SB500, it's the funky-shaped black rectangle one.  Today I got another SB500 working from another account (its E1: password worked) and it too reported a productSignature of 12.  I also know that when I was having problems with the config and the code was searching for slingboxes, the FinderID it reported matched (the one ending in 21c4).  I also remember his old account password being the decrypted password you reported, so maybe I'll try that in plaintext the next time I connect it.  I assume that even though the admin password came from a PRO-HD on the account it would still work on the SB500?  Or should I just be trying the user password for the SB500?

As I'm still a python noob and haven't isolated the decrypt routine yet, can you decrypt the user password for the box ending in 21c4 please?  I'll try that plaintext as well before I give up and reset it.

Now that I have multiple working boxes (a 350 and a 500 so far) I spent the evening working on the TiVo "network remote" code.  I made a config setting that - when set - intercepts the IR command and translates it to the TiVo network protocol command.  I got the intercept and translation working, now I just have to build the TCP socket to send the info.  I'm a total novice at threaded python sockets and was going cross-eyed trying to figure out where to insert the connection open, close, and socket_stillopen loop to keep the threading the rest of the script establishes (so I can do multiple sockets to multiple tivos).  Any suggestions would also be welcome.  I'll make another attempt on Monday.

Thanks for the help!

**Reply by GerryDazoo** [reply]
The slinger code prints out the cleartext of the encrypted password when it connects.
Slinginfo  "350/500/M1/M2" 12 30 10.21.0.89 5201 3072 10 **X7Ekfb32aIjQxZ2**

We might have a little bit of confusion on which password to use. There is the password to buddy's Sling account that is used to recover the individual passwords for the Slingboxes linked to that account. The sling account password will not work on the slingboxes.

Wondering where you hooked into the code for your Tivo? In the HTTP remote handling or later where to IR codes are sent to the slingbox? 

As far as the slingbox id I think we might be confusing data displayed in hex v.s. decimal   12 hex = 18 decimal.

**Reply by rbf121** [reply]
Good call on the hex, then there are 3 500s listed in your auth response.  Doesn't the slinger code list out the correct displayname on start up for all devices found on network?  does it match any in your list?

Also, I think I remember testing the userPassword but it might require code changes since by default it uses the "admin" username.  See code snippet here with 'admin':

        if not sling_cmd(0x67, pack('I 32s 32s 132x', 0, futf16('admin'), futf16(password))): # log in to SB
            print(name, 'Slingbox did not respond to login request. Please reset Slingbox and try again. NOT a factory reset') 
            return (s_ctl, None)
        if stat != 0:
            if stat == 0x2:
                print(name,'Error Starting Session. Check your admin password in config.ini file!')
            elif stat == 0x2b:
                print(name,'Error Starting Session. Slingbox might be Bricked')
            else:
                print(name,'Unknown Error Starting Session. Can''t Continue.')
            return s_ctl, None

**Reply by jsn081** [reply]
@rbf121 That's exactly what I was wondering about - if using the user password required a modification to the login code.  I hadn't found that line of code yet.  Thanks.  Unsure how I want to handle it, or if @GerryDazoo has a (more official) solution, as again I'm still a python hack.

@GerryDazoo I added an .ini setting called TivoTune, defaults to zero.  If it's set to 1 it intercepts the SendKeyCode and runs a translation to convert the code number to any available Tivo Network Remote command.  Most of the commands are simple, they are plaintext and are triggered by a CR.  So remote code 4 translates to `IRCODE CHANNELUP`.  After I actually test it I think I'm going to add one more chunk of code which calls for the TiVo to be forced into LiveTV mode before a channel change using the command `TELEPORT LIVETV`.  The reason is the behavior of a real TiVo and a TiVo Mini (I'm using both) is different when you fire random IR numbers at it.  Repeating the TELEPORT command has no ill effect, and is a good sanity check.

I'm now at the point where I need to establish the TCP connection, which I've specified in the .ini using TivoIP and TivoPort (which defaults to and should always be 31339).  But as I said, I'm not super familiar with threaded sockets.  The concept makes total sense, I'm just green with the nomenclature and the pointers.  I just need to check for the INI setting, on true import the IP and/or port, connect, and wait to send data.  I'm thinking about a once every 60s check of sending something invalid like `PING` and receiving back `INVALID_COMMAND` to verify the connection is up.

I'm happy to share what I've done so far.  Code below.

Line 699:
```
        
    def TivoRemoteSend( key ):
        TivoKeyList = [ 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'CHANNELUP' , 'CHANNELDOWN' , 'NULL' , 'NULL' , 'NULL' , 'NUM1' , 'NUM2' , 'NUM3' , 'NUM4' , 'NUM5' , 'NUM6' , 'NUM7' , 'NUM8' , 'NUM9' , 'NUM0' , 'ENTER' , 'NULL' , 'ENTER' , 'NULL' , 'NULL' , 'PLAY' , 'NULL' , 'PAUSE' , 'REVERSE' , 'FORWARD' , 'RECORD' , 'ADVANCE' , 'REPLAY' , 'LIVE' , 'NULL', 'NULL', 'GUIDE' , 'NULL' , 'CLEAR' , 'UP', 'DOWN' , 'LEFT' , 'RIGHT' , 'SELECT' , 'NULL' , 'NULL' , 'NULL' , 'INFO' , 'NULL' , 'NULL' , 'NULL' , 'SLOW' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'ZOOM' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'TIVO' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'NULL' , 'ACTION_C' , 'ACTION_D' , 'ACTION_A' , 'ACTION_B' ] 
        TivoKey = int(key)
        TivoCmd = 'IRCODE ' + TivoKeyList[TivoKey]
        print('Sending Tivo CMD: ', TivoCmd )
        # tcp command send goes here

    def SendKeycode( key, rccode ):

        UseTivo = slinginfo.get('TivoTune', 0 )
        if UseTivo == '1':
           print('Sending as Tivo command', key) #debugging
           TivoRemoteSend(key)
        else:
           print('sending key', key, rccode )
           if '.' in key:
...
```

I attempted to attach the protocol document but it failed, I can send it if you care to look.  My belief is that there's a new one they haven't made public (I have v1.1), because some of the functions don't work properly in the new TiVo GUI.  Even though there's FORCECH commands and whatnot, they tend to fail (even though theoretically they never should).  I've found the TELEPORT and IRCODE sets still operate reliably, so that's what I've used.

**Comment by jsn081** [comment]
> Doesn't the slinger code list out the correct displayname on start up for all devices found on network? does it match any in your list?

@rbf121 It does.  It matches the last one, ending in 21c4.  In a couple of hours I'm going to try hooking the box back up and sending the plaintext admin password.  If it doesn't and I want to try the user password, do we have a guess as to what the user is?  Hopefully 'user'? Lol.

**Reply by rbf121** [reply]
Pretty sure it's just "user" but if you have the admin password then you are probably better off just using that rather than changing code to support another username.

**Comment by jsn081** [comment]
So I tried a bunch of stuff, nothing worked.  I even changed the code to attempt to use "user" as the login, nothing worked.  So I'm guessing it's factory reset time?

In other news, I'm now less worried about the factory reset, because I modified the code to use the TiVo Network Remote.  I'm attaching the full script here as a proof-of-concept.  I'm not submitting it as any sort of feature request as I imagine the use-case is limited (as are my python skills - I had a lot of trial and error), but it seems to work really well.  I tried to use the word "Tivo" in every variable, so if you search the code for the word "Tivo" you should find all of the modifications (except for the `def is_socket_closed()` which I needed for sanity and I didn't see another procedure that did what I needed cleanly).  The override is triggered by setting `TivoTune = 1` in the box listing of the .ini file, and it requires `TivoIP = x.x.x.x` and optionally `TivoPort = 31339`.  It was tested with two boxes running simultaneously, using both the internal python remotes and external Plex integration using `Startchannel`.

(One note: In the remote code translation, I removed the "last" button and changed it to "back" - as my loaded TiVo remote IR doesn't have "back" IR code and the GUI desperately needs it.)

Both the protocol document and the modified script are below.  Maybe this will help someone else out on their project.

[TCP Remote Protocol, version 1.1.pdf](https://github.com/GerryDazoo/Slinger/files/14054543/TCP.Remote.Protocol.version.1.1.pdf)

[slingbox_server.py.txt](https://github.com/GerryDazoo/Slinger/files/14054533/slingbox_server.py.txt)
